runtime/lua/vim/inspect.lua

```diff
@@ -40,299 +53,289 @@ end
-local Inspector_mt = {__index = Inspector}
-
-function Inspector:puts(...)
-  local args   = {...}
-  local buffer = self.buffer
-  local len    = #buffer
-  for i=1, #args do
-    len = len + 1
-    buffer[len] = args[i]
-  end
+local function puts(buf, str)
+  buf.n = buf.n + 1
+  buf[buf.n] = str
 end
 
-function Inspector:down(f)
-  self.level = self.level + 1
-  f()
-  self.level = self.level - 1
-end
+local Inspector = {}
 
-function Inspector:tabify()
-  self:puts(self.newline, string.rep(self.indent, self.level))
-end
+local Inspector_mt = { __index = Inspector }
 
-function Inspector:alreadyVisited(v)
-  return self.ids[v] ~= nil
+local function tabify(inspector)
+  puts(inspector.buf, inspector.newline .. rep(inspector.indent, inspector.level))
 end
 
 function Inspector:getId(v)
   local id = self.ids[v]
+  local ids = self.ids
   if not id then
     local tv = type(v)
-    id              = (self.maxIds[tv] or 0) + 1
-    self.maxIds[tv] = id
-    self.ids[v]     = id
+    id = (ids[tv] or 0) + 1
+    ids[v], ids[tv] = id, id
   end
   return tostring(id)
 end
 
-function Inspector:putKey(k)
-  if isIdentifier(k) then return self:puts(k) end
-  self:puts("[")
-  self:putValue(k)
-  self:puts("]")
-end
+function Inspector:putValue(v)
+  local buf = self.buf
+  local tv = type(v)
+  if tv == "string" then
+    puts(buf, smartQuote(escape(v)))
+  elseif
+    tv == "number"
+    or tv == "boolean"
+    or tv == "nil"
+    or tv == "cdata"
+    or tv == "ctype"
+    or (vim and v == vim.NIL)
+  then
+    puts(buf, tostring(v))
+  elseif tv == "table" and not self.ids[v] then
+    local t = v
+
+    if t == inspect.KEY or t == inspect.METATABLE then
+      puts(buf, tostring(t))
+    elseif self.level >= self.depth then
+      puts(buf, "{...}")
+    else
+      if self.cycles[t] > 1 then
+        puts(buf, fmt("<%d>", self:getId(t)))
+      end
 
-function Inspector:putTable(t)
-  if t == inspect.KEY or t == inspect.METATABLE then
-    self:puts(tostring(t))
-  elseif self:alreadyVisited(t) then
-    self:puts('<table ', self:getId(t), '>')
-  elseif self.level >= self.depth then
-    self:puts('{...}')
-  else
-    if self.tableAppearances[t] > 1 then self:puts('<', self:getId(t), '>') end
-
-    local nonSequentialKeys, nonSequentialKeysLength, sequenceLength = getNonSequentialKeys(t)
-    local mt                = getmetatable(t)
-    if (vim and sequenceLength == 0 and nonSequentialKeysLength == 0
-        and mt == vim._empty_dict_mt) then
-      self:puts(tostring(t))
-      return
-    end
+      local keys, keysLen, seqLen = getKeys(t)
+      local mt = getmetatable(t)
 
-    self:puts('{')
-    self:down(function()
-      local count = 0
-      for i=1, sequenceLength do
-        if count > 0 then self:puts(',') end
-        self:puts(' ')
-        self:putValue(t[i])
-        count = count + 1
+      if vim and seqLen == 0 and keysLen == 0 and mt == vim._empty_dict_mt then
+        puts(buf, tostring(t))
+        return
```

Comment:
This bit from our local changes that is used to print `vim.empty_dict()` literally (instead of expanding the table) doesn't work yet. (Hence tests fail.)

Reply:
Looks fixed now \o/